include/cliffordsynthesis/CliffordSynthesizer.hpp

```diff
@@ -0,0 +1,289 @@
+    void setTableau(Tableau& targetTabl);
+    void optimize();
+
+    void setArchitecture(const Architecture& arch) {
+        architecture = arch;
+        if (nqubits == 0) {
+            nqubits = architecture.getNqubits();
+        }
+        auto map = highestFidelityMap.emplace_back();
+        architecture.getHighestFidelityCouplingMap(nqubits, map);
+    }
+    bool                   choose_best   = false;
+    bool                   use_embedding = false;
+    unsigned char          nqubits       = 0U;
+    std::set<signed char>  used_qubits{};
+    unsigned short         initial_timesteps = 0U;
+    int                    verbose           = 0;
+    int                    nthreads          = 1;
+    SynthesisStrategy      strategy          = SynthesisStrategy::UseMinimizer;
+    SynthesisTarget        target            = SynthesisTarget::GATES;
+    SynthesisMethod        method            = SynthesisMethod::Z3;
+    qc::QuantumComputation circuit;
+
+    std::vector<CouplingMap> highestFidelityMap;
+
+    Tableau initialTableau{};
+    Tableau targetTableau{};
+    Tableau modelTableau{};
+
+    void dumpResult(const std::string& outputFilename) {
+        if (optimal_results.resultCircuit.empty()) {
+            std::cerr << "Circuit is empty." << std::endl;
+            return;
+        }
+
+        size_t      dot       = outputFilename.find_last_of('.');
+        std::string extension = outputFilename.substr(dot + 1U);
+        std::transform(extension.begin(), extension.end(), extension.begin(), [](unsigned char c) { return ::tolower(c); });
+        if (extension == "real") {
+            dumpResult(outputFilename, qc::Real);
+        } else if (extension == "qasm") {
+            dumpResult(outputFilename, qc::OpenQASM);
+        }
+    }
+
+    void dumpResult(const std::string& outputFilename, qc::Format format) {
+        optimal_results.resultCircuit.dump(outputFilename, format);
+    }
+
+    void dumpResult(std::ostream& os, qc::Format format) {
+        optimal_results.resultCircuit.dump(os, format);
+    }
+
+    CliffordOptimizationResults optimal_results{};
+
+protected:
+    Architecture                architecture{};
+    CliffordOptimizationResults main_optimization(
+            int                                                        timesteps,
+            const std::set<std::pair<unsigned short, unsigned short>>& reducedCM,
+            const std::vector<unsigned short>& qubitChoice, Tableau& initialTableau,
+            Tableau& targetTableau);
+
+    void make_depth_optimizer(
+            int                                                        timesteps,
+            const std::set<std::pair<unsigned short, unsigned short>>& reducedCM,
+            const std::vector<unsigned short>& qubitChoice, std::unique_ptr<LogicBlock>& lb,
+            const LogicMatrix& x, const LogicMatrix& z, const LogicVector& r,
+            const LogicMatrix3D& g_s, const LogicMatrix3D& g_c) const;
```

Comment:
Would it make sense to capture all these logic variables in a `struct`. This would certainly allow for a cleaner interface here.